### PR TITLE
Mvbe 9586/handle django triggers with celery

### PR DESCRIPTION
**Pull request by berre94**
JIRA: https://vikingco.atlassian.net/browse/MVBE-9586

MVNE: https://github.com/vikingco/mvne-platform/pull/5579

### CODE REVIEW

File: djtriggers/logic.py [hunk 1]
```diff
@@ -147,3 +103,54 @@ def clean_triggers(expiration_dt=None, type_to_table=None):
...
+    Process given triggers asynchronously
+    :param django.db.models.query.QuerySet triggers: triggers to be processed
+    :param bool use_statsd: whether to use statsd
+    :return: None
+    """
+    logger.info('Start processing %d triggers of type %s asynchronously', triggers.count(), triggers[0].trigger_type)
+
+    group([trigger.process.subtask({'use_statsd': use_statsd},
+                                   max_retries=getattr(settings, 'DJTRIGGERS_CELERY_TASK_MAX_RETRIES', 0))
+           for trigger in triggers]).apply_async()
```
Comment (Duologic) [inline djtriggers/logic.py]: I'm not sure how this works, is it possible to write this logic to be more readable so it is more clear what it does?

File: djtriggers/migrations/0005_auto_20160804_1706.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('djtriggers', '0004_trigger_nb_of_tries'),
+    ]
+
+    operations = [
+        migrations.AlterField(
```
Comment (Duologic) [inline djtriggers/migrations/0005_auto_20160804_1706.py]: Would it be possible to squash these 2 migrations into a single one?

File: djtriggers/models.py [hunk 3]
```diff
@@ -82,28 +90,95 @@ def set_source(self, *args):
...
+        :param dict dictionary: dictionary needed by trigger to execute
+        :param bool use_statsd: whether to use statsd
+        :return: None
+        """
+        # The task gets locked because multiple tasks in the queue can process the same trigger.
+        # The lock assures no two tasks can process a trigger simultaneously.
+        # The check for date_processed assures a trigger is not executed multiple times.
+        try:
+            trigger_lock_name = 'execute_trigger%d' % self.pk
+            file_lock = FileLock(trigger_lock_name)
```
Comment (Duologic) [inline djtriggers/models.py]: FileLock is not an option I think, as Celery runs on multiple servers at once. We should lock it in the database. In mvne-platform, we use https://github.com/vikingco/django-db-locking, beware of the different branches here, @nielsvanoch should be able to help you with that.

Comment (nielsvanoch) [inline djtriggers/models.py]: Indeed, we should use db-locking for that. An example: https://github.com/vikingco/mvne-platform/blob/master/mvne/telco/balance_checker/tasks.py#L100

File: djtriggers/models.py [hunk 4]
```diff
@@ -82,28 +90,95 @@ def set_source(self, *args):
...
 
+    def _handle_execution_failure(self, exception, use_statsd=False):
+        """
+        Handle execution failure of the trigger
+        :param Exception exception: the exception raised during failure
+        :param bool use_statsd: whether to use statsd
+        :return: None
+        """
+        self.nb_of_tries += 1
+        if self.nb_of_tries >= getattr(settings, 'DJTRIGGERS_TRIGGER_RETRIES_WARNING_LVL', sys.maxint):
```
Comment (Duologic) [inline djtriggers/models.py]: What is the default value for `DJTRIGGERS_TRIGGER_RETRIES_WARNING_LVL`?
As this value compares to `nb_of_tries`, I think the name of the settings can be a bit more clear. Something like `DJTRIGGERS_WARNING_AFTER_RETRIES` or something. `LVL` sounds like you can define the logging level here (ie. INFO, WARNING, ERROR, ...).

Also, I have a feeling that using abbreviations like `nb` and `LVL` is not very useful. The cost of typing `number` and `LEVEL` is fairly low compared to the readability. Avoid abbreviations as much as possible, it doesn't make your code faster.

Comment (Duologic) [inline djtriggers/models.py]: Here is a good read from @kristofclaes on readability: https://kristofclaes.github.io/2016/07/15/code-should-communicate-intent/ It talks about avoiding abbreviations and reducing line count.

File: djtriggers/models.py [hunk 5]
```diff
@@ -59,9 +66,10 @@ class Trigger(models.Model):
 
     trigger_type = models.CharField(max_length=50, db_index=True)
     source = models.CharField(max_length=250, null=True, blank=True, db_index=True)
-    date_received = models.DateTimeField(default=datetime.now)
+    date_received = models.DateTimeField(default=timezone.now)
     date_processed = models.DateTimeField(null=True, blank=True, db_index=True)
     process_after = models.DateTimeField(null=True, blank=True, db_index=True)
+    nb_of_tries = models.IntegerField(default=0)
```
Comment (Duologic) [inline djtriggers/models.py]: Avoid abbreviations.

File: djtriggers/models.py (hunk 3 above)
Comment (berre94) [inline djtriggers/models.py]: @Duologic Okay, but django-db-locking does not support 'non-waiting' locks. So every trigger will be waiting until the lock is released, after which it will raise an 'AlreadyProcessedError'. This is the reason I switched to FileLock which allows you to proceed immediately if the trigger is locked.

File: djtriggers/models.py [hunk 6]
```diff
@@ -82,28 +90,95 @@ def set_source(self, *args):
...
-            self.logger.log_result(self, self._process(dictionary))
-        except ProcessLaterError as e:
-            self.process_after = e.process_after
-            self.save()
-            raise
-        if self.date_processed is None:
-            self.date_processed = now
-        self.save()
+            if logger:
+                self.logger = get_logger(logger)
```
Comment (Duologic) [inline djtriggers/models.py]: I know this is how it worked before, maybe check with a logging-guru if it wouldn't be better to define the 'logger' outside the class (global) and refer to that?

File: djtriggers/tests/test_models.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
+from datetime import timedelta
+from logging import WARNING
+
+import mock
+import pytest
```
Comment (Duologic) [inline djtriggers/tests/test_models.py]: Can you add requirement files for both regular install and test? 

Have a look here on how to accomplish this:
https://github.com/vikingco/cl_payments/tree/master/deployment
https://github.com/vikingco/cl_payments/blob/master/setup.py (look at the for loops over the files)

File: djtriggers/models.py (hunk 3 above)
Comment (Duologic) [inline djtriggers/models.py]: A FileLock doesn't work when running on multiple servers, so that is not an option. Please have discussion with your fellow Dev's on how to solve that 'non-waiting' problem.

**Comment by WimCrols** [comment]
Can you rebase your commits into fewer, logical commits? And handfull should suffice.

Also take care of making your commit messages descriptive. A message such as 'renaming' viewed in 6 months will be not be clear enough.

**Comment by WimCrols** [comment]
Please provide a description in your Pull Request linking back to the JIRA ticket.

**Comment by WimCrols** [comment]
Isn't there a mvne-platform Pull Request too?

**Comment by berre94** [comment]
@WimCrols yeah sorry, i forgot. here it is: https://github.com/vikingco/mvne-platform/pull/5579

File: djtriggers/logic.py [hunk 8]
```diff
@@ -147,3 +102,58 @@ def clean_triggers(expiration_dt=None, type_to_table=None):
...
+    count_done, count_error, count_exception = 0, 0, 0
+
+    # Process each trigger
+    for trigger in triggers:
+        try:
+            trigger.process.apply(use_statsd=use_statsd)
+        # The trigger didn't need processing yet
+        except (ProcessLaterError, AlreadyProcessedError):
+            pass
+        # The trigger raised an (expected) error while processing
```
Comment (WimCrols) [inline djtriggers/logic.py]: Why is there a difference between ProcessError exception and others? What kind of errors are "expected" then?

File: djtriggers/loggers/database.py [hunk 9]
```diff
@@ -1,18 +1,26 @@
+import logging
+
 from djtriggers.loggers.base import TriggerLogger
 
 class DatabaseLogger(TriggerLogger):
-    def log_result(self, trigger, message):
+    def log_result(self, trigger, message, level=None):
```
Comment (Duologic) [inline djtriggers/loggers/database.py]: `level` is not used in this function

File: djtriggers/loggers/database.py [hunk 10]
```diff
@@ -1,18 +1,26 @@
...
+
 def _prettify(results):
     try:
         return '\n'.join([str(r) for r in results])
     except TypeError:
         return str(results)
 
 class DatabaseSerializeLogger(DatabaseLogger):
-    def log_result(self, trigger, results):
+    def log_result(self, trigger, results, level=None):
```
Comment (Duologic) [inline djtriggers/loggers/database.py]: Same here, `level` is not used in this function.

File: djtriggers/tests/test_models.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+    def test_handle_execution_failure(self):
+        exception = Exception()
+        trigger = NormalDummyTriggerFactory()
+        original_tries = trigger.number_of_tries
+        trigger._handle_execution_failure(exception)
+
+        assert trigger.number_of_tries == original_tries + 1
+
+    @override_settings(DJTRIGGERS_TRIGGER_RETRIES_WARNING_LVL=3)
```
Comment (Duologic) [inline djtriggers/tests/test_models.py]: Isn't this renamed to `DJTRIGGERS_TRIES_BEFORE_WARNING` ?

File: requirements.txt [hunk 12]
```diff
@@ -0,0 +1,6 @@
+celery>=3.1
+Django>=1.8,<1.10
+python-dateutil
+
+# Locking of objects using a database table
+-e git+ssh://git@github.com/vikingco/django-db-locking.git@1.1.3#egg=django-db-locking
```
Comment (Duologic) [inline requirements.txt]: You can drop the `-e` here, some users of this library might not have a Github account.

File: setup.py [hunk 13]
```diff
@@ -21,3 +48,4 @@
         'Framework :: Django',
     ],
 )
+
```
Comment (Duologic) [inline setup.py]: Not sure what the policy is about empty newlines at the end of the file, personally I prefer not doing that.

File: djtriggers/tests/test_models.py (hunk 7 above)
Comment (Duologic) [inline djtriggers/tests/test_models.py]: This comment is covered.

File: djtriggers/logic.py (hunk 8 above)
Comment (berre94) [inline djtriggers/logic.py]: This was existing code. I have not encountered any usage of this error in the code of django-triggers. My suspicion is that, while writing the trigger._process(), this exception can be used to denote exceptions that are expected to happen within execution of the trigger.

File: djtriggers/loggers/database.py (hunk 9 above)
Comment (berre94) [inline djtriggers/loggers/database.py]: redundant argument is needed because method is overriden from TriggerLogger

File: djtriggers/loggers/database.py (hunk 10 above)
Comment (berre94) [inline djtriggers/loggers/database.py]: redundant argument is needed because method is overriden from TriggerLogger

File: djtriggers/tests/models.py [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+
+class FailingDummyTrigger(Trigger):
+    class Meta:
+        proxy = True
+
+    typed = 'failing_trigger_test'
+
+    def _process(self, dictionary):
+        raise Exception
```
Comment (WimCrols) [inline djtriggers/tests/models.py]: If you raise constructor on line 34, then also do it here please.

File: djtriggers/tests/models.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    class Meta:
+        proxy = True
+
+    typed = 'normal_trigger_test'
+
+    def _process(self, dictionary):
+        pass
+
+
+class FailingDummyTrigger(Trigger):
```
Comment (WimCrols) [inline djtriggers/tests/models.py]: I would call this RaisesExceptionDummyTrigger for clarity.

File: djtriggers/tests/models.py [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    class Meta:
+        proxy = True
+
+    typed = 'failing_trigger_test'
+
+    def _process(self, dictionary):
+        raise Exception
+
+
+class ProcessErrorDummyTrigger(Trigger):
```
Comment (WimCrols) [inline djtriggers/tests/models.py]: And then also here RaisesProcessErrorDummyTrigger.

File: djtriggers/tests/models.py [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    class Meta:
+        proxy = True
+
+    typed = 'process_error_normal_trigger_test'
+
+    def _process(self, dictionary):
+        raise ProcessError()
+
+
+class TimelyDummyTrigger(Trigger):
```
Comment (WimCrols) [inline djtriggers/tests/models.py]: Timely? That means it's on time, while the code shows it is slower. Isn't that misleading?

File: djtriggers/tests/test_logic.py [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+from djtriggers.tests.factories.triggers import NormalDummyTriggerFactory, FailingDummyTriggerFactory
+
+
+class LogicTest(TestCase):
+
+    @staticmethod
+    def populate_db(quantity, factory, **kwargs):
+        return [factory(**kwargs) for _ in xrange(quantity)]
+
+    def setUp(self):
```
Comment (WimCrols) [inline djtriggers/tests/test_logic.py]: Remove this code since it does nothing.

File: djtriggers/tests/test_logic.py [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+from django.test.testcases import TestCase
+from django.utils import timezone
+
+from djtriggers.logic import process_triggers
+from djtriggers.tests.factories.triggers import NormalDummyTriggerFactory, FailingDummyTriggerFactory
+
+
+class LogicTest(TestCase):
+
+    @staticmethod
```
Comment (WimCrols) [inline djtriggers/tests/test_logic.py]: You define it as static, but then later on use it as an instance method. Then I would just remove the decorator and be done with it.

File: djtriggers/tests/test_logic.py [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+        normal_triggers_due2 = self.populate_db(7, NormalDummyTriggerFactory,
+                         process_after=now - timedelta(days=1))
+        self.populate_db(11, NormalDummyTriggerFactory,
+                         process_after=now + timedelta(days=1))
+        self.populate_db(13, NormalDummyTriggerFactory,
+                         date_processed=now)
+        process_triggers()
+
+        normal_triggers_due.extend(normal_triggers_due2)
+        for call in mock_process_sync.call_args_list:
```
Comment (WimCrols) [inline djtriggers/tests/test_logic.py]: I don't understand this part. Note to self to ask for more explanation.

File: djtriggers/tests/test_models.py [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+from djtriggers.loggers.base import TriggerLogger
+from djtriggers.models import Trigger
+from djtriggers.tests.factories.triggers import NormalDummyTriggerFactory, TimelyDummyTriggerFactory, \
+    FailingDummyTriggerFactory
+from djtriggers.tests.models import NormalDummyTrigger, FailingDummyTrigger, TimelyDummyTrigger
+
+
+class TriggerTest(TestCase):
+    """
+    Due to some bug in celery.contrib.methods.task_method testing the tash method needs be done in the following way:
```
Comment (WimCrols) [inline djtriggers/tests/test_models.py]: typo: tash

File: djtriggers/tests/test_models.py [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+        Class.TaskMethod(class_instance)
+    """
+    def setUp(self):
```
Comment (WimCrols) [inline djtriggers/tests/test_models.py]: Please remove.

File: djtriggers/tests/test_models.py [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+
+    def test_handle_execution_success(self):
+        trigger = NormalDummyTriggerFactory()
+        trigger._handle_execution_success()
+
+        assert trigger.date_processed
+        assert timezone.now() - trigger.date_processed < timedelta(minutes=1)
+
+    @mock.patch('django_statsd.clients.statsd')
+    def test_test_handle_execution_success_use_statsd(self, mock_statsd):
```
Comment (WimCrols) [inline djtriggers/tests/test_models.py]: test_test ? Probably a copy-paste error?

File: djtriggers/tests/test_models.py [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+        trigger._handle_execution_success(use_statsd=True)
+
+        assert trigger.date_processed
+        assert timezone.now() - trigger.date_processed < timedelta(minutes=1)
+        mock_statsd.incr.assert_called_with('triggers.%s.processed' % trigger.trigger_type)
+        mock_statsd.timing.assert_called_with('triggers.%s.process_delay_seconds' % trigger.trigger_type,
+                              (trigger.date_processed - trigger.process_after).total_seconds())
+
+    def test_handle_execution_failure(self):
+        exception = Exception()
```
Comment (WimCrols) [inline djtriggers/tests/test_models.py]: Is there any reason you instantiate Exception() beforehand in these couple of test cases?

File: djtriggers/models.py [hunk 25]
```diff
@@ -82,28 +89,88 @@ def set_source(self, *args):
...
+    def _handle_execution_failure(self, exception, use_statsd=False):
+        """
+        Handle execution failure of the trigger
+        :param Exception exception: the exception raised during failure
+        :param bool use_statsd: whether to use statsd
+        :return: None
+        """
+        self.number_of_tries += 1
+        if self.number_of_tries >= getattr(settings, 'DJTRIGGERS_TRIES_BEFORE_WARNING', 3):
+            message = 'Processing of %s %s raised a %s (try nb. %s) ' % (self.trigger_type,
```
Comment (WimCrols) [inline djtriggers/models.py]: "number" gets abbreviated as "nr." or "#". Or you can just write "try 3".

File: djtriggers/tests/test_models.py [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+    @override_settings(DJTRIGGERS_TRIES_BEFORE_WARNING=3)
+    @mock.patch.object(TriggerLogger, 'log_message')
+    def test_handle_execution_failure_tries_exceeded(self, mock_logger):
+        exception = Exception()
+        trigger = NormalDummyTriggerFactory(number_of_tries=3)
+        original_tries = trigger.number_of_tries
+        trigger._handle_execution_failure(exception)
+
+        assert trigger.number_of_tries == original_tries + 1
+        message = 'Processing of %s %s raised a %s (try nb. %s) ' % (trigger.trigger_type,
```
Comment (WimCrols) [inline djtriggers/tests/test_models.py]: Please prefer _format()_ function over _%_.

File: djtriggers/tests/test_models.py [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+
+    @mock.patch('django_statsd.clients.statsd')
+    def test_handle_execution_failure_use_statsd(self, mock_statsd):
+        exception = Exception()
+        trigger = NormalDummyTriggerFactory()
+        original_tries = trigger.number_of_tries
+        trigger._handle_execution_failure(exception, use_statsd=True)
+
+        assert trigger.number_of_tries == original_tries + 1
+        mock_statsd.incr.assert_called_with('triggers.%s.failed' % trigger.trigger_type)
```
Comment (WimCrols) [inline djtriggers/tests/test_models.py]: Please prefer _format()_ function over _%_.

File: djtriggers/logic.py [hunk 28]
```diff
@@ -147,3 +102,58 @@ def clean_triggers(expiration_dt=None, type_to_table=None):
...
+        except ProcessError:
+            count_error += 1
+        # A general exception occurred
+        except:
+            count_exception += 1
+        # The trigger was successfully processed
+        else:
+            count_done += 1
+
+    logger.info('success: %s, error: %s, exception: %s' % (count_done, count_error, count_exception))
```
Comment (WimCrols) [inline djtriggers/logic.py]: Please prefer _format()_ function over _%_.

File: djtriggers/models.py (hunk 25 above)
Comment (WimCrols) [inline djtriggers/models.py]: Please prefer _format()_ function over _%_.

File: djtriggers/models.py [hunk 29]
```diff
@@ -82,28 +89,88 @@ def set_source(self, *args):
...
+                                                                         self.pk,
+                                                                         type(exception).__name__,
+                                                                         self.number_of_tries)
+            self.logger.log_message(self, message, level=WARNING)
+
+        # Send stats to statsd if necessary
+        if use_statsd:
+            from django_statsd.clients import statsd
+            statsd.incr('triggers.%s.failed' % self.trigger_type)
```
Comment (WimCrols) [inline djtriggers/models.py]: Please prefer _format()_ function over _%_.

File: djtriggers/models.py [hunk 30]
```diff
@@ -82,28 +89,88 @@ def set_source(self, *args):
...
+        :return: None
+        """
+        now = timezone.now()
+        if self.date_processed is None:
+            self.date_processed = now
+
+        # Send stats to statsd if necessary
+        if use_statsd:
+            from django_statsd.clients import statsd
+            statsd.incr('triggers.%s.processed' % self.trigger_type)
```
Comment (WimCrols) [inline djtriggers/models.py]: Please prefer _format()_ function over _%_.

File: djtriggers/tests/test_models.py [hunk 31]
```diff
@@ -0,0 +1,108 @@
...
+
+    @mock.patch('django_statsd.clients.statsd')
+    def test_test_handle_execution_success_use_statsd(self, mock_statsd):
+        trigger = NormalDummyTriggerFactory(process_after=timezone.now())
+        trigger._handle_execution_success(use_statsd=True)
+
+        assert trigger.date_processed
+        assert timezone.now() - trigger.date_processed < timedelta(minutes=1)
+        mock_statsd.incr.assert_called_with('triggers.%s.processed' % trigger.trigger_type)
+        mock_statsd.timing.assert_called_with('triggers.%s.process_delay_seconds' % trigger.trigger_type,
```
Comment (WimCrols) [inline djtriggers/tests/test_models.py]: Please prefer _format()_ function over _%_.

File: djtriggers/tests/test_models.py (hunk 24 above)
Comment (berre94) [inline djtriggers/tests/test_models.py]: In the testcase without 'use_statsd' I don't need it, in the testcase with 'use_statsd' I need it to construct the message to assert.

File: djtriggers/models.py (hunk 25 above)
Comment (berre94) [inline djtriggers/models.py]: always or just when a lot of '%' occur?

File: djtriggers/models.py [hunk 32]
```diff
@@ -82,31 +89,93 @@ def set_source(self, *args):
...
+                    self._handle_execution_success(use_statsd)
+                except ProcessLaterError as e:
+                    self.process_after = e.process_after
+                    self.save()
+                    raise
+                except Exception as e:
+                    self._handle_execution_failure(e, use_statsd)
+                    raise
+        except AlreadyLocked:
+            pass
```
Comment (WimCrols) [inline djtriggers/models.py]: If a trigger is already locked, shouldn't it be retried? Or does that happen automatically?

Comment (Duologic) [inline djtriggers/models.py]: This task is run every X minutes, so if locked but didn't execute/complete, it will be retried by this job:
https://github.com/vikingco/django-triggers/blob/master/djtriggers/tasks.py#L7
Currently, the task is run with a management command, but we should convert that to a periodic task as described in the ticket.

**Comment by WimCrols** [comment]
:shipit:

**Comment by KMahoney** [comment]
[LGTM](http://lgtm.in/i/2leFe8mpq)
:shipit:

File: djtriggers/models.py [hunk 33]
```diff
@@ -75,35 +82,111 @@ def __init__(self, *args, **kwargs):
...
+                if not force and self.date_processed is not None:
+                    raise AlreadyProcessedError()
+                if not force and self.process_after and self.process_after >= now:
+                    raise ProcessLaterError(self.process_after)
+
+                try:
+                    # execute trigger
+                    self.logger.log_result(self, self._process(dictionary))
+                    self._handle_execution_success(use_statsd)
+                except ProcessLaterError as e:
```
Comment (RoelandMatthijssens) [inline djtriggers/models.py]: i'm not entirely sure when and how this may occur. We already check if we should process this trigger later and if not we execute the trigger. How can it still raise a processLater error?

File: djtriggers/models.py [hunk 34]
```diff
@@ -75,35 +82,111 @@ def __init__(self, *args, **kwargs):
...
 
+    def _handle_execution_failure(self, exception, use_statsd=False):
+        """
+        Handle execution failure of the trigger
+        :param Exception exception: the exception raised during failure
+        :param bool use_statsd: whether to use statsd
+        :return: None
+        """
+        self.number_of_tries += 1
+        if self.number_of_tries >= getattr(settings, 'DJTRIGGERS_TRIES_BEFORE_WARNING', 3):
```
Comment (RoelandMatthijssens) [inline djtriggers/models.py]: there is an off by one here since we increment before checking and include equals in the check. if `tries_before_warning = 3` it will only be tried twice.

File: djtriggers/tests/factories/triggers.py [hunk 35]
```diff
@@ -0,0 +1,53 @@
...
+
+    trigger_type = 'failing_trigger_test'
+    source = 'tests'
+    date_received = timezone.now()
+    date_processed = None
+    process_after = None
+    number_of_tries = 0
+
+
+class RaisesProcessErrorDummyTriggerFactory(DjangoModelFactory):
```
Comment (RoelandMatthijssens) [inline djtriggers/tests/factories/triggers.py]: Im not sure what the intention of these error factories is. This one is not used in tests and afaik for the other one we could just as easily mock the process method to throw an error in the tests that require the trigger to fail

File: djtriggers/tests/test_logic.py [hunk 36]
```diff
@@ -0,0 +1,45 @@
...
+from django.utils import timezone
+
+from djtriggers.logic import process_triggers
+from djtriggers.tests.factories.triggers import NormalDummyTriggerFactory
+
+
+class LogicTest(TestCase):
+
+    def populate_db(self, quantity, factory, **kwargs):
+        return [factory(**kwargs) for _ in xrange(quantity)]
```
Comment (RoelandMatthijssens) [inline djtriggers/tests/test_logic.py]: I like this approach :) but for clarity i'd make the quantity a named field (probably personal preference though)

File: djtriggers/tests/test_models.py [hunk 37]
```diff
@@ -0,0 +1,106 @@
...
+    Due to some bug in celery.contrib.methods.task_method testing the task method needs be done in the following way:
+        Class.TaskMethod(class_instance)
+    """
+
+    def test_handle_execution_success(self):
+        trigger = NormalDummyTriggerFactory()
+        trigger._handle_execution_success()
+
+        assert trigger.date_processed
+        assert timezone.now() - trigger.date_processed < timedelta(minutes=1)
```
Comment (RoelandMatthijssens) [inline djtriggers/tests/test_models.py]: I'd prefer calling the trigger inside a fake_time block so we can make an exact assertion of the processed date

Comment (RoelandMatthijssens) [inline djtriggers/tests/test_models.py]: This occurs a couple of times in the tests. Same applies for every occurance

File: djtriggers/tests/test_models.py [hunk 38]
```diff
@@ -0,0 +1,106 @@
...
+    def test_handle_execution_failure(self):
+        trigger = NormalDummyTriggerFactory()
+        original_tries = trigger.number_of_tries
+        trigger._handle_execution_failure(Exception())
+
+        assert trigger.number_of_tries == original_tries + 1
+
+    @override_settings(DJTRIGGERS_TRIES_BEFORE_WARNING=3)
+    @mock.patch.object(TriggerLogger, 'log_message')
+    def test_handle_execution_failure_tries_exceeded(self, mock_logger):
```
Comment (RoelandMatthijssens) [inline djtriggers/tests/test_models.py]: the edge case of the happy path is not tested number_of_tries = 2 should not raise an exception (currently this is not the case)

File: djtriggers/logic.py [hunk 39]
```diff
@@ -129,21 +71,64 @@ def clean_triggers(expiration_dt=None, type_to_table=None):
...
+            process_trigger(trigger, use_statsd=use_statsd)
+        # The trigger didn't need processing yet
+        except (ProcessLaterError, AlreadyProcessedError):
+            pass
+        # The trigger raised an (expected) error while processing
+        except ProcessError:
+            count_error += 1
+        # A general exception occurred
+        except:
+            count_exception += 1
```
Comment (nielsvanoch) [inline djtriggers/logic.py]: Fair enough, will add logging.

File: djtriggers/models.py (hunk 33 above)
Comment (nielsvanoch) [inline djtriggers/models.py]: The `_process` function of the trigger itself may raise it, if it decided that the `process_after` date was incorrect and it wants to be executed later.

File: djtriggers/models.py (hunk 34 above)
Comment (nielsvanoch) [inline djtriggers/models.py]: Will update.

File: djtriggers/tests/factories/triggers.py (hunk 35 above)
Comment (nielsvanoch) [inline djtriggers/tests/factories/triggers.py]: Agreed. Will refactor.

File: djtriggers/tests/test_logic.py (hunk 36 above)
Comment (nielsvanoch) [inline djtriggers/tests/test_logic.py]: Will update.

File: djtriggers/tests/test_models.py (hunk 37 above)
Comment (nielsvanoch) [inline djtriggers/tests/test_models.py]: Looking at it, I think the check actually adds very little value. I'm removing them completely.

File: djtriggers/tests/test_models.py (hunk 38 above)
Comment (nielsvanoch) [inline djtriggers/tests/test_models.py]: Will add a test.

**Comment by nielsvanoch** [comment]
@RoelandMatthijssens updated. Recheck?

File: djtriggers/tests/test_models.py [hunk 40]
```diff
@@ -0,0 +1,110 @@
...
+    def test_handle_execution_failure(self):
+        trigger = NormalDummyTriggerFactory()
+        original_tries = trigger.number_of_tries
+        trigger._handle_execution_failure(Exception())
+
+        assert trigger.number_of_tries == original_tries + 1
+
+    @override_settings(DJTRIGGERS_TRIES_BEFORE_WARNING=3)
+    @mock.patch.object(TriggerLogger, 'log_message')
+    def test_handle_execution_failure_success_after_one_failure(self, mock_logger):
```
Comment (RoelandMatthijssens) [inline djtriggers/tests/test_models.py]: Test function name contains failure while te test is not testing failures

**Comment by nielsvanoch** [comment]
Updated.

**Comment by nielsvanoch** [comment]
Closing in favor of https://github.com/vikingco/django-triggers/pull/16
